src/golang/lib/engine/aq_engine.go

```diff
@@ -846,14 +884,22 @@ func (eng *aqEngine) execute(
 	}
 
 	if len(completedOps) != len(dag.Operators()) {
-		return errors.Newf("Internal error: %d operators were provided but only %d completed.", len(dag.Operators()), len(completedOps))
+		err = errors.Newf("Internal error: %d operators were provided but only %d completed.", len(dag.Operators()), len(completedOps))
+		return err
 	}
 
 	for opID, depCount := range opToDependencyCount {
 		if depCount != 0 {
-			return errors.Newf("Internal error: operator %s has a non-zero dep count %d.", opID, depCount)
+			err = errors.Newf("Internal error: operator %s has a non-zero dep count %d.", opID, depCount)
+			return err
 		}
 	}
+
+	// avoid overriding a warning notification.
+	if notificationLevel != mdl_shared.WarningNotificationLevel {
+		hasNotification = true
+		notificationLevel = mdl_shared.SuccessNotificationLevel
```

Comment:
Another advantage of packaging the level and msg between a struct is that we can write a set() method that enforces notification level logic like this outside of the core orchestration code.

Reply:
Also, this if-statement is true for ErrorNotificationLevels too right? We wouldn't want to override that either right.

Reply:
Since the next line is `return nil` we should safely assume everything is `error free` and no need to worry about the error level case. Do you feel anything is not self-explanatory here?

Reply:
I think what I was confused about was that in the error case, we are expected to return an `opFailureError()` before getting here. In that case, can we assert here that the notificationLevel cannot be ERROR?

Reply:
actually I see what you mean, yeah having the check covering 'ErrorNotfLevel' makes sense. It won't be the case in practice but this code block doesn't need to make this assumption